src/main/java/org/opensearch/neuralsearch/processor/factory/SemanticFieldProcessorFactory.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.processor.factory;
+
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.env.Environment;
+import org.opensearch.index.analysis.AnalysisRegistry;
+import org.opensearch.ingest.AbstractBatchingSystemProcessor;
+import org.opensearch.neuralsearch.util.SemanticMappingUtils;
+import org.opensearch.neuralsearch.ml.MLCommonsClientAccessor;
+import org.opensearch.neuralsearch.processor.semantic.SemanticFieldProcessor;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.opensearch.plugins.IngestPlugin.SystemIngestPipelineConfigKeys.INDEX_MAPPINGS;
+import static org.opensearch.plugins.IngestPlugin.SystemIngestPipelineConfigKeys.INDEX_TEMPLATE_MAPPINGS;
+
+/**
+ *  Factory for semantic fields.
+ *
+ *  This factory is for internal usage and will be systematically invoked if we detect there are semantic fields
+ *  defined. Users should not be able to define this type of the processor in a regular ingest pipeline.
+ */
+public final class SemanticFieldProcessorFactory extends AbstractBatchingSystemProcessor.Factory {
+    public static final String PROCESSOR_FACTORY_TYPE = "system_ingest_processor_semantic_field";
+    /**
+     * Ideally this should be configurable through the semantic fields. For P0 we will use a default value,
+     * and in P1 we will make it configurable.
+     * TODO: Make batch size configurable
+     */
+    private static final int DEFAULT_BATCH_SIZE = 10;
+
+    private final MLCommonsClientAccessor mlClientAccessor;
+
+    private final Environment environment;
+
+    private final ClusterService clusterService;
+    private final AnalysisRegistry analysisRegistry;
+
+    public SemanticFieldProcessorFactory(
+        final MLCommonsClientAccessor mlClientAccessor,
+        final Environment environment,
+        final ClusterService clusterService,
+        final AnalysisRegistry analysisRegistry
+    ) {
+        super(PROCESSOR_FACTORY_TYPE);
+        this.mlClientAccessor = mlClientAccessor;
+        this.environment = environment;
+        this.clusterService = clusterService;
+        this.analysisRegistry = analysisRegistry;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected AbstractBatchingSystemProcessor newProcessor(String tag, String description, Map<String, Object> config) {
+        final List<Map<String, Object>> mappings = new ArrayList<>();
+        final Object mappingFromIndex = config.get(INDEX_MAPPINGS);
+        final Object mappingFromTemplates = config.get(INDEX_TEMPLATE_MAPPINGS);
+        if (mappingFromTemplates instanceof List) {
+            mappings.addAll((List<Map<String, Object>>) mappingFromTemplates);
+        }
+        if (mappingFromIndex instanceof Map) {
+            mappings.add((Map<String, Object>) mappingFromIndex);
+        }
```

Comment:
Since we try to execute and construct system ingest pipelines for every index request, is this additional object creation overhead for every single index request too? Should we move the validations higher up?

Reply:
nvm, this is mitigated by system ingest pipeline caching in core